### PR TITLE
Automated host listing its own errata

### LINKED ISSUE #3950: Host can't list its own errata

### PR DESCRIPTION
**Pull request by renzon**
close #3950

### CODE REVIEW

**Comment by renzon** [comment]
Test result indicating the error:
```console
>           raise CLIReturnCodeError(*error_data)
E           CLIReturnCodeError: CLIReturnCodeError(return_code=70, stderr=u"[ERROR 2016-10-27 09:53:46 Exception] undefined method `installable_errata' for nil:NilClass\nundefined method `installable_errata' for nil:NilClass\n[ERROR 2016-10-27 09:53:46 Exception] \n\nRestClient::InternalServerError (500 Internal Server Error):\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/abstract_response.rb:48:in `return!'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:308:in `block in rest_client_call_block'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:228:in `call'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:228:in `process_result'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:178:in `block in transmit'\n    /opt/rh/rh-ruby22/root/usr/share/ruby/net/http.rb:853:in `start'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:172:in `transmit'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:64:in `execute'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:33:in `execute'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/resource.rb:51:in `get'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:295:in `call_client'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:224:in `http_call'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:174:in `call_action'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:169:in `call'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/resource.rb:14:in `call'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/apipie/command.rb:43:in `send_request'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:190:in `send_request'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:280:in `send_request'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:315:in `retrieve_all'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:299:in `execute'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:133:in `run'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/bin/hammer:125:in `<top (required)>'\n    /usr/bin/hammer:23:in `load'\n    /usr/bin/hammer:23:in `<main>'\n\n\n", msg=u'Command "host errata list" finished with return_code 70\nstderr contains following message:\n[ERROR 2016-10-27 09:53:46 Exception] undefined method `installable_errata\' for nil:NilClass\nundefined method `installable_errata\' for nil:NilClass\n[ERROR 2016-10-27 09:53:46 Exception] \n\nRestClient::InternalServerError (500 Internal Server Error):\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/abstract_response.rb:48:in `return!\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:308:in `block in rest_client_call_block\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:228:in `call\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:228:in `process_result\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:178:in `block in transmit\'\n    /opt/rh/rh-ruby22/root/usr/share/ruby/net/http.rb:853:in `start\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:172:in `transmit\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:64:in `execute\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/request.rb:33:in `execute\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/rest-client-1.6.7/lib/restclient/resource.rb:51:in `get\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:295:in `call_client\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:224:in `http_call\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:174:in `call_action\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/api.rb:169:in `call\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/apipie-bindings-0.0.16/lib/apipie_bindings/resource.rb:14:in `call\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/apipie/command.rb:43:in `send_request\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:190:in `send_request\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:280:in `send_request\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:315:in `retrieve_all\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.5.1.9/lib/hammer_cli_foreman/commands.rb:299:in `execute\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/lib/hammer_cli/abstract.rb:22:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:133:in `run\'\n    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.5.1.11/bin/hammer:125:in `<top (required)>\'\n    /usr/bin/hammer:23:in `load\'\n    /usr/bin/hammer:23:in `<main>\'\n\n\n'

../../../robottelo/cli/base.py:162: CLIReturnCodeError
============================= 45 tests deselected ==============================
=================== 1 failed, 45 deselected in 5.19 seconds ====================
```

**Comment by renzon** [comment]
Tests for 6.3 ok:

```console
/home/renzo/PycharmProjects/venvs/robottelo/bin/python /home/renzo/bin/pycharm-2016.2/helpers/pycharm/pytestrunner.py -p pytest_teamcity /home/renzo/PycharmProjects/robottelo/tests/foreman/cli/test_host.py "-k HostErrataTestCase and test_positive_host_itself_errata_listing"
Testing started at 2:56 PM ...
============================= test session starts ==============================
=================== 1 passed, 45 deselected in 7.63 seconds ====================

**Comment by abalakh** [comment]
Could you please point to BZ or something what requested such test? For me it doesn't look like functional test, more like unit test as you aren't working with any errata but basically just verifying that "something" was returned when the command was called.
I'd rather request the test which creates a host (or uses some existing one), installs some old package to that host and checks that some specific errata which will update that specific package is listed by `errata_list` command.

File: tests/foreman/cli/test_host.py [hunk 1]
```diff
@@ -1316,3 +1316,19 @@ def test_negative_unregister_and_pull_content(self):
...
+    def test_positive_host_itself_errata_listing(self):
+        """Check if can list errata from the host itself
+
+        @id: 6b22f0c0-9c4b-11e6-ab93-68f72889dc7f
+
+        @assert: Host itself can list errata
+
+        @BZ: 1351040
+        """
+        self.assertIsInstance(Host.errata_list({'host-id': 1}), list)
```
Comment (pondrejk) [inline tests/foreman/cli/test_host.py]: Is it certain that the host running the server has always id 1? Should we use the hostname instead, or determine the id from the hostname on the fly? 

Also, I don't think this scenario will work by default, as the host needs to be self-registered to see it's own errata. I didn't try that but downstream docs suggest so: https://access.redhat.com/documentation/en/red-hat-satellite/6.2/paged/server-administration-guide/chapter-3-configuring-a-self-registered-satellite

File: tests/foreman/cli/test_host.py [hunk 2]
```diff
@@ -1316,3 +1316,19 @@ def test_negative_unregister_and_pull_content(self):
         result = self.client.run(
             'yum install -y {0}'.format(FAKE_1_CUSTOM_PACKAGE))
         self.assertNotEqual(result.return_code, 0)
+
+
+class HostErrataTestCase(CLITestCase):
+    """Tests for errata's host sub command"""
+
+    @tier1
+    def test_positive_host_itself_errata_listing(self):
```
Comment (sghai) [inline tests/foreman/cli/test_host.py]: `test_positive_errata_listing_of_sat_server` ?

File: tests/foreman/cli/test_host.py [hunk 3]
```diff
@@ -1316,3 +1316,19 @@ def test_negative_unregister_and_pull_content(self):
...
+        """Check if can list errata from the host itself
```
Comment (sghai) [inline tests/foreman/cli/test_host.py]: Check if errata can be listed for a Satellite/Foreman server

**Comment by sghai** [comment]
If we really want to automate the [bz](https://bugzilla.redhat.com/show_bug.cgi?id=1351040) then we just need to validate if `host errata list --host-id=<server_id>` doesn't throw any exception.

self registering the satellite is not a good idea, specially with automation-boxes.

@renzon : I see the bz is verified. So what do you get with `hammer -u admin -p changeme host errata list  --host-id 1`

**Comment by omaciel** [comment]
@sghai @pondrejk can you guys drive this PR forward?

**Comment by renzon** [comment]
@abalakh BZ is on comment: https://bugzilla.redhat.com/show_bug.cgi?id=1351040. In fact it doesn't work at all, even if no errata is set up. @sghai I got exception on 6.2:

```console
[root@foo ~]# rpm -q satellite
satellite-6.2.4-1.0.el7sat.noarch
[root@cloud-qe-15 ~]# hostname
foo.redhat.com
[root@foo~]#
[root@foo ~]# hammer shell
Welcome to the hammer interactive shell
Type 'help' for usage information
hammer> host list
---|------------------------------------------|------------------|------------|--------------|------------------
ID | NAME                                     | OPERATING SYSTEM | HOST GROUP | IP           | MAC              
---|------------------------------------------|------------------|------------|--------------|------------------
1  | foo.redhat.com | RHEL Server 7.3  |            | IP Omitted | Mac Omitted
---|------------------------------------------|------------------|------------|--------------|------------------
hammer> host errata list --host-id 1
undefined method `installable_errata' for nil:NilClass
```

But it worked on 6.3:

```console

```console
[root@bar ~]# rpm -q satellite
satellite-6.3.0-6.1.beta.el7sat.noarch
[root@cloud-qe-15 ~]# hostname
bar.redhat.com
[root@bar~]#
[root@bar ~]# hammer shell
Welcome to the hammer interactive shell
Type 'help' for usage information
hammer> host list
---|------------------------------------------|------------------|------------|--------------|------------------
ID | NAME                                     | OPERATING SYSTEM | HOST GROUP | IP           | MAC              
---|------------------------------------------|------------------|------------|--------------|------------------
1  | foo.redhat.com | RHEL Server 7.3  |            | IP Omitted | Mac Omitted
---|------------------------------------------|------------------|------------|--------------|------------------
hammer> host errata list --host-id 1
hammer>
```
```
That is how I verified and the followed same steps on automation.

@pondrejk gut questions. On both systems the host itself got id 1. But you are right, better not trust on that, I am changing it to fetch by hostname.

**Comment by renzon** [comment]
I just reset and erased an 6.3 box and when got nothing when listing host, I thought that Satellite woud list itself as host on installation, it is not the case? @sghai @omaciel @pondrejk

**Comment by renzon** [comment]
The session on the box:

```console
hammer> host errata list --host-id 1
ActiveRecord::RecordNotFound
```

**Comment by sghai** [comment]
@renzon:  As I said before, its a bit difficult to automate the test by properly populating errata for server itself just because we may need to self register the satellite and once it is done we can't sync packages from CDN.

So I would vote to just check when you run `host errata list --host-id 1` that you won't get any exception/error. Self registering Satellite and related functions are completely comes under a different feature.

And lets discuss this once offline too for better understanding.

**Comment by pondrejk** [comment]
@renzon -- I think you are hitting the difference betw. *host* and *content host*.  Content host is registered to Sat and runs katello-agent which allows you to manage errata. My understanding is that Sat lists itself as a host by default but to make it content host you have to register it to itself. In Sat 6.2 cli, you can compare `hammer host list` with `hammer content-host list`, but the second command is no longer available in 6.3 :/

**Comment by renzon** [comment]
@pondrejk the bz is specific about the command error, it is host: https://bugzilla.redhat.com/show_bug.cgi?id=1351040. 

The funny think is that today I got no error in box. E.g., I was able to see the satellite itself on host list, My only explanation for it is that when I run the the command on my last command, I did it just after installation and maybe there was a background task to include sat server itself as host.

I talked to @sghai and there is a completely different use case for self registering a server and you can not do that now because doing so would make several tests fail because after registration satellite would stop to get updates from CDN.

So if you satellite itself is listed on host list by default, I think this test is good to go. I made some changes on test name and explicitly mentioning BZ field so you guys can check details.

**Comment by sghai** [comment]
ACK

**Pull request merged**